src/util/runtimeloggingcategory.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <QByteArray>
+#include <QLoggingCategory>
+
+/// RuntimeLoggingCategory is a wrapper around QLoggingCategory
+/// that facilitates using a QLoggingCategory for a category
+/// name string that is determined at run time. To use a
+/// RuntimeLoggingCategory, simply pass it to the
+/// qCInfo, qCDebug, qCWarning, or qCCritical macros like a
+/// QLoggingCategory. If the category name is fully known at
+/// compile time, simply construct a QLoggingCategory with a
+/// string literal instead of using RuntimeLoggingCategory.
+///
+/// RuntimeLoggingCategory stores the logging category name as a
+/// QByteArray. This is helpful because QLoggingCategory has a deleted
+/// copy constructor, does not take ownership nor make a copy of the
+/// const char* passed to its constructor, and requires that the
+/// data pointed to by the const char* lives as long as the
+/// QLoggingCategory.
+///
+/// RuntimeLoggingCategory takes advantage of QByteArray's
+/// implicit sharing to ensure that the const char* for the
+/// QLoggingCategory lives as long as the QLoggingCategory without
+/// requiring excessive deep copies.
+struct RuntimeLoggingCategory {
+  public:
+    RuntimeLoggingCategory(const QByteArray& oldByteArray)
+            : byteArray(oldByteArray),
+              logger(byteArray.constData()) {
+    }
+
+    RuntimeLoggingCategory(const RuntimeLoggingCategory& oldCategory)
```

Comment:
I would prefer to make the expensive *cloning* operation more explicit instead of hiding it behind an innocent copy ctor.

Is this possible without any copy/move ctor?

```
static RuntimeLoggingCategory clone(const RuntimeLoggingCategory& other) {
    return RuntimeLoggingCategory(other.m_category);
}
```

Reply:
How is this expensive? [QByteArray's copy constructor](https://doc.qt.io/qt-5/qbytearray.html#QByteArray) uses implicit sharing.

Reply:
It's about semantics, not performance. Pass by value and move into place for taking ownership.

Reply:
Like [this](https://github.com/mixxxdj/mixxx/pull/4239/commits/d63269da2dfb73429194e70bc1ee9cf3e274a774)? I don't really get the point, but okay.

Reply:
If you pass-by-value the caller can decide to either give up ownership by using std::move on their side or by passing in a copy and retaining ownership. If you pass-by-reference the copy ctor will be invoked regardless of the intention of the caller, because you **have** to copy it.

Reply:
Just a convention. If you only need to *borrow* temporarily for the duration of the call use pass-by-const-ref otherwise use pass-by-value to take take ownership.

Reply:
I don't understand. If I change the constructors of the classes that take a RuntimeLoggingCategory to take a `const RuntimeLoggingCategory` instead of a `const RuntimeLoggingCategory&`, initializing the `const RuntimeLoggingCategory` members fails to compile because the copy constructor is private.

Reply:
Unfortunately, the Core Guidelines now recommend to pass via `&&` for consuming parameters, which is less convenient for the caller:

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-consume

Reply:
Changing the parameter type to `const RuntimeLoggingCategory&&` still tries to invoke the copy constructor which is private.

Reply:
... unless you want the class members to be `const RuntimeLoggingCategory&&` too instead of `const RuntimeLoggingCategory`?

Reply:
If it is not possible in C++ then let's forget about it.

Alternative option: Define a custom constructor that takes an empty marker `struct Clone {}` as 2nd unused parameter and otherwise works exactly as the current copy constructor. That could not be invoked accidentally.

Reply:
> let's forget about it

I agree. I just removed the last commit making the copy constructor private. Let's not try too hard to bend C++ into something that it is not.